graphql-server/src/schemas/custom.ts

```diff
@@ -37,8 +43,69 @@ import {
 } from "./queries/requests";
 import { typeDefs } from "../utils/typeDefs";
 const request = require("request-promise-native");
+import { AuthenticationError, ForbiddenError } from "apollo-server-express";
+import { applyMiddleware } from "graphql-middleware";
+import { IMiddlewareResolver } from "graphql-middleware/dist/types";
+import { ExecuteStatementOptions } from "@databricks/sql/dist/contracts/IDBSQLSession";
+import { queryDatabricks } from "../utils/databricks";
+
+const KEYCLOAK_PHI_ACCESS_GROUP = "mrn-search";
+
+type AuthMiddleware = {
+  Query: {
+    dashboardPatients: IMiddlewareResolver;
+  };
+};
 
 export async function buildCustomSchema(ogm: OGM) {
+  const authenticationMiddleware: AuthMiddleware = {
+    Query: {
+      dashboardPatients: async (
+        resolve,
+        parent,
+        args: QueryDashboardPatientsArgs,
+        context: ApolloServerContext,
+        info
+      ) => {
+        if (
+          args.phiEnabled &&
+          args.searchVals &&
+          args.searchVals.length > 0 &&
+          !context.req.isAuthenticated()
+        ) {
+          throw new AuthenticationError(
+            "You must be logged in to access this resource."
+          );
+        }
+        return await resolve(parent, args, context, info);
+      },
+    },
+  };
+
+  const authorizationMiddleware: AuthMiddleware = {
+    Query: {
+      dashboardPatients: async (
+        resolve,
+        parent,
+        args: QueryDashboardPatientsArgs,
+        context: ApolloServerContext,
+        info
+      ) => {
+        if (
+          args.phiEnabled &&
+          args.searchVals &&
+          args.searchVals.length > 0 &&
+          !context.req.user.groups.includes(KEYCLOAK_PHI_ACCESS_GROUP)
+        ) {
+          throw new ForbiddenError(
+            "You do not have permission to access this resource. Please contact the SMILE team for assistance."
+          );
+        }
+        return await resolve(parent, args, context, info);
+      },
+    },
+  };
```

Comment:
Added middlewares to inspect requests and ensure that users making patients queries asking for PHI data are properly authenticated and authorized. This middleware setup and logic were mostly copied over from the previous `GetPatientIdsTriplets` query that fetches from the CRDB.

Reply:
Does authentication have a timeout?

Reply:
Yes. Users are logged out 10m after being idle. This setting is done in the Keycloak admin.